### PR TITLE
Let migrations not explode memory

### PR DESCRIPTION
**Pull request by gerolf-da**
This is a slightly spicy one, as it touches existing scala DB migrations.

This doesn't cause migration hash conflicts in flyway for existing databases, because java/scala migrations don't have an associated hash.

When running a recent version of daml on sql on a database from last August 2020, I noticed two issues:
1. ValueSerializer ran into a match error for values that weren't deprecated (fixed by 6675887).
2. The database had a bit over 2 million rows in `participant_events`. Without setting the fetch size explicitly, the JDBC driver (un)happily tries to load the entire result set into memory, eventually crashing the participant with an OutOfMemoryError (fixed by 075326a).

I initially suspected the memory explosion in the V38 migration to come from the assembly of the `LazyList` value (`Stream` often caused similar issues by holding on to values too long), so I changed it to `Iterator`. I then kept the `Iterator` based approach, as reasoning about what is kept in memory is easier with `Iterator` than with `Stream`/`LazyList` imho.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by SamirTalwar-DA — APPROVED** [review]
This looks pretty good to me. I haven't checked that you caught all of them; I figure your large database dump would have helped with that.

File: ledger/participant-integration-api/src/main/scala/db/migration/postgres/V32_1__Fix_key_hashes.scala [hunk 1]
```diff
@@ -25,8 +27,11 @@ private[migration] final class V32_1__Fix_key_hashes extends BaseJavaMigration {
       fixHash = conn.prepareStatement(FIX_HASH)
 
       selectKeys = conn.createStatement()
+      conn.setFetchSize(BATCH_SIZE)
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/db/migration/postgres/V32_1__Fix_key_hashes.scala:30]: I think you meant to use `selectKeys.setFetchSize` here.

Comment (gerolf-da) [reply ledger/participant-integration-api/src/main/scala/db/migration/postgres/V32_1__Fix_key_hashes.scala:30]: yeah, thanks. 🤦

File: ledger/participant-integration-api/src/main/scala/db/migration/translation/ValueSerializer.scala [hunk 2]
```diff
@@ -39,6 +39,7 @@ private[migration] object ValueSerializer {
         logger.error(
           s"*** Please upgrade your sandbox database by upgrading your SDK to 1.6 first. ***"
         )
+      case _ =>
```
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/db/migration/translation/ValueSerializer.scala:42]: @remyhaemmerle-da: When running this migration on an index DB from last August, this match failed on the `Right` side. I only could get it to pass after add this wildcard match. Would this match your expectation of how this should work?

Comment (remyhaemmerle-da) [reply ledger/participant-integration-api/src/main/scala/db/migration/translation/ValueSerializer.scala:42]: That is the only think that make sens.  This should need to be backported, shouldn't  it ?

Comment (remyhaemmerle-da) [reply ledger/participant-integration-api/src/main/scala/db/migration/translation/ValueSerializer.scala:42]: I propose to simplify this unnecessary complexity of mine:

```scala

  private[this] val DeprecatedValueVersions = Set("1", "2", "3", "4", "5")
  private[this] val UnsupportedErrorMessage = """Unsupported transaction version '(\d)'""".r

  private[translation] def handleDeprecatedValueVersions[X](x: Either[ValueCoder.DecodeError, X]): Either[ValueCoder.DecodeError, X] = {
    x match {
      case Left(ValueCoder.DecodeError(UnsupportedErrorMessage(version))) if DeprecatedValueVersions(version) =>
        logger.error(
          s"*** Deserialization of value version $version is not supported by the SDK 1.7.0 or later. ***"
        )
        logger.error(
          s"*** Please upgrade your sandbox database by upgrading your SDK to 1.6 first. ***"
        )
      case _ => 
    }
    x
  }
```

**Pull request merged**